pkg/channel/consolidated/reconciler/controller/consumer_group_watcher.go

```diff
@@ -0,0 +1,197 @@
+type ConsumerGroupWatcher interface {
+	// Start instructs the watcher to start polling for the consumer groups and
+	// notify any observers on the event of any changes
+	Start() error
+
+	// Terminate instructs the watcher to stop polling and clear the watchers cache
+	Terminate()
+
+	// Watch registers callback on the event of any changes observed
+	// on the consumer groups. watcherID is an arbitrary string the user provides
+	// that will be used to identify his callbacks when provided to Forget(watcherID).
+	// Successive calls with the same watcherID will queue the callbacks successfully.
+	//
+	//To ensure this is event-triggered, level-driven,
+	// we don't pass the updates to the callback, instead the observer is expected
+	// to use List() to get the updated list of ConsumerGroups.
+	Watch(watcherID string, callback ConsumerGroupHandler) error
+
+	// Forget removes all callbacks that correspond to the watcherID
+	Forget(watcherID string)
+
+	// List returns all the cached consumer groups that match matcher.
+	// It will return an empty slice if none matched or the cache is empty
+	List(matcher Matcher) []string
+}
+
+type WatcherImpl struct {
+	logger *zap.SugaredLogger
+	//TODO name?
+	watchers             map[string][]ConsumerGroupHandler
+	cachedConsumerGroups []string
+	adminClient          kafka.AdminClient
+	pollDuration         time.Duration
+	done                 chan struct{}
+}
+
+func NewConsumerGroupWatcher(ctx context.Context, ac kafka.AdminClient, pollDuration time.Duration) ConsumerGroupWatcher {
+	return &WatcherImpl{
+		logger:               logging.FromContext(ctx),
+		adminClient:          ac,
+		pollDuration:         pollDuration,
+		watchers:             make(map[string][]ConsumerGroupHandler),
+		cachedConsumerGroups: make([]string, 0),
+	}
+}
+
+func (w *WatcherImpl) Start() error {
+	w.logger.Infow("ConsumerGroupWatcher starting. Polling for consumer groups", zap.Duration("poll duration", w.pollDuration))
+	go func() {
+		for {
+			select {
+			case <-after(w.pollDuration):
+				// let's get current observed consumer groups
+				observedCGs, err := w.adminClient.ListConsumerGroups()
+				if err != nil {
+					w.logger.Errorw("error while listing consumer groups", zap.Error(err))
+					continue
+				}
+				var notify bool
+				var changedGroup string
+				// Look for observed CGs
+				for _, c := range observedCGs {
+					if !Find(w.cachedConsumerGroups, c) {
+						// This is the first appearance.
+						w.logger.Debugw("Consumer group observed. Caching.",
+							zap.String("consumer group", c))
+						changedGroup = c
+						notify = true
+						break
+					}
+				}
+				// Look for disappeared CGs
+				for _, c := range w.cachedConsumerGroups {
+					if !Find(observedCGs, c) {
+						// This CG was cached but it's no longer there.
+						w.logger.Debugw("Consumer group deleted.",
+							zap.String("consumer group", c))
+						changedGroup = c
+						notify = true
+						break
+					}
+				}
+				if notify {
```

Comment:
If a consumer group is coming from a different application, notify (in `Start()`) will always be true, so every watcher will be called. (since `cg` in notify(cg) isn't used.)
I think we need a mechanism to exclude observed consumer groups that are not part of our application.

Reply:
so, initially I designed this in a typical observer pattern where `notify` watch/notify can notify a subset of watchers giving them which objects added/deleted based on them watching for specific consumer groups. But that was more of a pure edge-triggered observer. To make it more "level driven" similar to k8s reconciliation, I stopped providing "what" changed, and instead notified of a change, where it's the responsibility of the observer to pull current state via `List` and a matcher.

That said, maybe we can have a filtered watch, `Watch(watcherID, watcher, callback)` where observers provide a matcher, and only modifications passing those would get notified. As long as it is still level driven by not passing what changes and forcing them to query for current state of the world I'm fine by that approach. What do you think? @pierDipi

Reply:
Ok, thanks for the clarification. 

> That said, maybe we can have a filtered watch, Watch(watcherID, watcher, callback) where observers provide a matcher, and only modifications passing those would get notified. As long as it is still level driven by not passing what changes and forcing them to query for current state of the world I'm fine by that approach. What do you think?

It makes sense to me, it's mostly an optimization, so I think we can do it later as long as we keep track of it.